File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.py
class Solution:
    def findLCS(self, word1: str, word2: str) -> int:
        m = len(word1)
        n = len(word2)
        maxLen = 0
        dp = [0] * (n + 1)
        for i in range(1, m + 1):
            current = [0] * (n + 1)
            for j in range(1, n + 1):
                # match
                if word1[i - 1] == word2[j - 1]:
                    current[j] = 1 + dp[j - 1]
                else:
                    current[j] = max(dp[j], current[j - 1])
            dp = current
        return dp[n]
    
    def minDistance(self, word1: str, word2: str) -> int:
        return len(word1) + len(word2) - 2 * self.findLCS(word1, word2)

File: 583-delete-operation-for-two-strings/583-delete-operation-for-two-strings.cpp
class Solution {
private:
    int findLCS(string &text1, string &text2) {
        int n1 = text1.size(), n2 = text2.size();
        vector<int> dp(n2 + 1, 0);
        for(int idx1 = 1; idx1 <= n1; ++idx1) {
            vector<int> current(n2 + 1, 0);
            for(int idx2 = 1; idx2 <= n2; ++idx2) {
                // match
                if(text1[idx1 - 1] == text2[idx2 - 1]) 
                    current[idx2] = 1 + dp[idx2 - 1];
                else 
                    current[idx2] = max(dp[idx2], current[idx2 - 1]);
            }
            dp = current;
        }
        return dp[n2];
    }
public:
    int minDistance(string word1, string word2) {
        if(word1.size() == word2.size())
            return 2 * (word1.size() - findLCS(word1, word2));
        return word1.size() + word2.size() - 2 * findLCS(word1, word2);
    }
};